dawn/src/dawn/AST/Offsets.h

```diff
@@ -100,11 +117,14 @@ class HorizontalOffset {
 
 template <typename T>
 T offset_cast(HorizontalOffset const& offset) {
-  return dynamic_cast<T>(*offset.impl_);
+  static std::remove_reference_t<T> nullOffset{};
```

Comment:
I see a problem here: when you have no `impl_` and the user asked for a reference such as `T=CartesianOffset&`, this function will return a reference to a static object (`nullOffset`), which is not what we want, as the user will likely modify it.

Reply:
A solution would be to detect whether `T` is a reference and do something like :
```
// if T is a reference {
   if(!offset.impl_) {
      offset.impl_ = std::make_unique<std::remove_reference_t<T>>();
      return dynamic_cast<T>(*offset.impl_);
   }
// }
```

Reply:
Mhh he shouldn't be able to call this with `T=CartesianOffset&`. The idea is that the user cannot modify the offset through `offset_cast` (that's ok because this function is only needed for CodeGen. Your version doesn't work because you cannot return a valid reference to that object; it will go out of scope immediately.  I think currently it would be possible to cast to non-const (because the pointer is non-const) even though the signature suggests that it's not possible; but I add a static_assert.

Reply:
What would go out of scope? If you `make_unique` your implementation and then return a reference to it, this will persist, because you hold it as a member.
But if we don't need it, it's ok to disallow `T` being a reference... in case, instead, we realize we could need it, let's just keep in mind that it is still possible